lib/ansible/modules/files/file.py

```diff
@@ -433,16 +445,21 @@ def ensure_directory(path, follow, recurse):
             raise AnsibleModuleError(results={'msg': 'There was an issue creating %s as requested:'
                                                      ' %s' % (curpath, to_native(e)),
                                               'path': path})
+        return {'path': path, 'changed': changed, 'diff': diff}
 
-    # We already know prev_state is not 'absent', therefore it exists in some form.
     elif prev_state != 'directory':
+        # We already know prev_state is not 'absent', therefore it exists in some form.
         raise AnsibleModuleError(results={'msg': '%s already exists as a %s' % (path, prev_state),
                                           'path': path})
 
+    #
```

Comment:
why not an `else` or `elif prev_state == 'directory'` around all this to avoid the repetition of the return? That's a maintenance trip hazard...

Reply:
I can't remember if it was Jack Dietrich's talk here at pycon or Armin Ronacher's slide deck about API design but returning early should make this less of a trip hazard.  When you return early, you know at the end of the conditional block that you don't have to read any further in the function for more information and don't have to jump back and forth to look for what a variable's value might be.

Once I added the return value to the condtition for that reason, it seemed to make sense to remove the ```elif prev_state == 'directory':``` which I had before.  Removing it makes the code read as having one main expected path (the idempotent path in this case), and several branches from it.  It also means that you don't end the code with a single return value (where, if you're having to trace back through the code from the return, you then have to figure out where the condition jumped from.)  I could see putting the code (sans final return) back into an ```elif``` block but after thinking about it for a few days,  I think it makes the code less clear to do it that way.